src/jaxsim/rbda/contacts/soft.py

```diff
@@ -157,179 +200,161 @@ class SoftContacts(ContactModel):
-                        f_tangential / jnp.maximum(jnp.linalg.norm(f_tangential), 1e-9)
-                    )
-
-                    # Sum the normal and tangential forces, and create the 6D force.
-                    CW_f_slip = force_normal + f_tangential_projected
-                    CW_f = jnp.hstack([CW_f_slip, jnp.zeros(3)])
-
-                    # Correct the material deformation derivative for slipping contacts.
-                    # Basically we compute ṁ such that we get `f_tangential` on the cone
-                    # given the current (m, δ).
-                    ε = 1e-9
-                    δε = jnp.maximum(δ, ε)
-                    α = -K * jnp.sqrt(δε)
-                    β = -D * jnp.sqrt(δε)
-                    ṁ = (f_tangential_projected - α * m) / β
-
-                    # Return the 6D force in the contact frame and
-                    # the deformation derivative.
-                    return CW_f, ṁ
-
-                CW_f, ṁ = jax.lax.cond(
-                    pred=f_tangential.dot(f_tangential) > (μ * force_normal_mag) ** 2,
-                    true_fun=lambda _: slipping_contact(),
-                    false_fun=lambda _: sticking_contact(),
-                    operand=None,
-                )
-
-                # Express the 6D force in the world frame.
-                W_f = W_Xf_CW @ CW_f
-
-                # Return the 6D force in the world frame and the deformation derivative.
-                return W_f, (ṁ,)
-
-            # (W_f, (ṁ,))
-            return jax.lax.cond(
-                pred=active_contact,
-                true_fun=lambda _: below_terrain(),
-                false_fun=lambda _: above_terrain(),
-                operand=None,
-            )
+        return W_Xf_CW @ CW_f, (ṁ,)
 
-        # (W_f, (ṁ,))
-        return jax.lax.cond(
-            pred=(μ == 0.0),
-            true_fun=lambda _: with_no_friction(),
-            false_fun=lambda _: with_friction(),
-            operand=None,
-        )
+    @staticmethod
+    @jax.jit
+    def compute_penetration_data(
+        p: jtp.VectorLike,
+        v: jtp.VectorLike,
+        terrain: jaxsim.terrain.Terrain,
+    ) -> tuple[jtp.Float, jtp.Float, jtp.Vector]:
+
+        # Pre-process the position and the linear velocity of the collidable point.
+        W_ṗ_C = jnp.array(v).squeeze()
+        px, py, pz = jnp.array(p).squeeze()
+
+        # Compute the terrain normal and the contact depth.
+        n̂ = terrain.normal(x=px, y=py).squeeze()
+        h = jnp.array([0, 0, terrain.height(x=px, y=py) - pz])
+
+        # Compute the penetration depth normal to the terrain.
+        δ = jnp.maximum(0.0, jnp.dot(h, n̂))
+
+        # Compute the penetration normal velocity.
+        δ̇ = -jnp.dot(W_ṗ_C, n̂)
+
+        # Enforce the penetration rate to be zero when the penetration depth is zero.
+        δ̇ = jnp.where(δ > 0, δ̇, 0.0)
+
+        return δ, δ̇, n̂
```

Comment:
Since we are using this in various parts of the code, we might consider putting it in `jaxsim.api.contact`